### DISCUSSION TITLE
how to add annotation to monaco editor in theia

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-theia/theia#13784] [Q&A]
hi,
    1. I use monaco editor to add some annotation to c++ method out of theia, for example use follow code to add an "line comment " to c++ method
![image](https://github.com/eclipse-theia/theia/assets/4537809/3207f0c5-fb0f-4399-880a-683e25761234)
![image](https://github.com/eclipse-theia/theia/assets/4537809/7822fdc1-683b-4a3e-86f6-a9c85309cf72)
    2. now I want to implement above in my theia extension, but I do not find function like registerCodeLensProvider to in theia's monaca editor, I am not familiar it. Could any one give some idea about how to implement it in theia monaco editor

**Comment by msujew** [comment]
Hey @wss29,

the recommended way of adding code lenses to Theia is by writing a VSCode extension that provides the feature and including that into your Theia App as a builtin extension.

Aside from that, you're also free to dig through our own implementation of this, but I wouldn't recommend it, since the internal Monaco/Theia API is not stable, while the VSCode API will always be stable.

**Comment by wss29** [comment]
Hi msujew,
    thanks for your advice, vscode indeed has some inline APIsupport codelens well, like registerCodeLensProvider. and it will save much time, but I still have some problems if write codelens by vscode extension 
    1. does it mean I can not use theia's inline API like reactwidget、command and so on? This will lose the advantage of flexibility in theia
    2. if I have some backend service or frontend widget UI written in theia extension, vscode extension will not be able to use them?
    3. When developing codelens through vscode extension, can theia support it well?

**Reply by msujew** [reply]
Note that you could still create your own code lens provider in code, but you'll need to dig into this, see [this related code](https://github.com/eclipse-theia/theia/blob/d79b7101adef9db71e74c13744a3fde4f79a343c/packages/plugin-ext/src/main/browser/languages-main.ts#L512-L546). Just because I don't recommend to do it, doesn't mean it's not technically feasible.

> 1. does it mean I can not use theia's inline API like reactwidget、command and so on? This will lose the advantage of flexibility in theia

You can, just because you additionally build a vscode plugin doesn't mean you suddenly cannot build a Theia extension. You use VS Code for the code lens, and a Theia extension for everything else.

> 2. if I have some backend service or frontend widget UI written in theia extension, vscode extension will not be able to use them?

See answer above. Note that you can always communicate between a VS Code plugin and the Theia frontend using the commands API.

> 3. When developing codelens through vscode extension, can theia support it well?

Yes? I wouldn't recommend to use it otherwise.

**Reply by wss29** [reply]
Hi msujew:
    thanks for your patient replay, 
    I originally thought vscode extensions were an isolated module for theia, I have developed many theia extensions, and if can not communicate with vscode externsion, it will be bad, and I will lose many resources and redevelop them in vscode. 
    now that know vscode extension can communicate with theia, I decided to follow your advice to use vscode extension for codelens and theia extension for others. where can I find simple code about communicating between vscode extension and theia extension? I want to learn it, any simple will save me time,  thanks again

**Reply by wss29** [reply]
@msujew  Hi msujew:
I have implemented some codelens in vscode extensions, but I do not know how to communicate between a VS Code plugin and the Theia extension, can you give me some advice about where can I find a sample solution?

**Reply by msujew** [reply]
You can implement message passing via commands. You can implement a command on the VS Code plugin/Theia extension side (probably both) and invoke them via the `CommandRegistry` or `vscode.commands.executeCommand` API respectively. Since commands can accept arbitrary arguments and can return arbitrary values, you can use them to implement an RPC system from the Theia frontend to the VS Code plugin.
